man/rom41cat.1
.TH rom41cat 1 06-November-2024 "LIF Utilities" "LIF Utilities"
.SH NAME
rom41cat \- display the function names in an HP41 ROM image
.SH SYNOPSIS
.B lifutils rom41cat
[\-x][\-a]
<
.I Input file
[>
.I Output_file
]
.PP
.B lifutils rom41cat
[\-x][\-a]
.I INPUTFILE
[>
.I Output_file
]
.PP
.B lifutils rom41cat \-?
.SH DESCRIPTION
.B rom41cat
reads the Function Address Table (FAT) at the start of an unscrambled 
HP41 ROM image and displays the names of the functions (both Mcode and 
user language (Focal)) contained in that ROM. Input is read either from standard input
or the file specified by the command line parameter
.I INPUTFILE.
.PP
If a ROM consists of multiple 4K ROM files, they must be concatenated before
feeding them into
.B rom41cat
The maximum size of concatenated ROM files is 16K. See the example below.
.PP
If the 
.B \-x
option is not given, then the XROM numbers, entry address of the function, 
language, and name are displayed on standard output. 
.PP
If the 
.B \-x
option is given, then the XROM numbers and function name are displayed on 
standard output in a format that can be stored in a file and read into 
.B comp41
as an XROM definition file. Then 
.B comp41
can give the correct names to functions in this ROM. In this case, the 
names of user language functions are displayed in the form
.I XROM'name'
.PP
Use 
.B er41rom
to unscramble Eramco ROM image files before using
.B rom41cat
to display the function names contained therein.
.SH OPTIONS
.TP
.I \-x
Produce a
.B comp41
xrom definition file
.TP
.I \-a
Decompile the following non ASCII HP-41 FOCAL characters as Unicode characters: text append (hex 0x251c),
GREEK small mu (hex 0x03bc), angle (hex 0x2280), not equal to (hex 0x2260), GREEK capital 
sigma (hex 0x03A3).  In a XROM file function names containing these FOCAL characters are output as two 
text strings. The first string contains these characters as escaped hex codes, the second as Unicode characters. 
Note: XROM files created with this option are compatible with
.B comp41
, 
.B decomp41
, 
.B keys41
but not with other software that support xrom files.
.TP 
.I \-?
Print a message giving program usage to standard error.
.SH REFERENCES
The format of the FAT and names for Mcode functions is given in 
.I HP\-41 Mcode for Beginners (Ken Emery)
. The format of names for user language functions does not appear to be 
documented and was discovered by experiment.
.SH EXAMPLES
If 
.I myrom.rom
is an Eramco ROM image file, then
.PP
.B lifutils er41rom < myrom.rom | lifutils rom41cat
.PP
will display the function names, entry points, and language to standard 
output, while
.PP
.B lifutils er41rom < myrom.rom | lifutils rom41cat -x > myrom.xrom
.PP
will produce an XROM definition file for
.B comp41
containing the function names in 
.I myrom.rom
.PP
To create an XROM file for the Advantage Pac, which consists of 3 ROM image files:
.PP
.B cat AdvL1-1B.rom AdvU1-1B.rom AdvU2-1B.rom | lifutils rom41cat -x -a > advantage.xrom (Unix, macOS)
.PP
or
.PP
.B copy /b AdvL1-1B.rom+AdvU1-1B.rom+AdvU2-1B.rom | lifutils rom41cat -x -a > advantage.xrom (Windows)
.SH BUGS
Some commonly used HP41 characters have no ASCII equivalents and are 
therefore displayed as \\xnn hex escape sequences. Spaces in function 
names are replaced by underscores since 
.B comp41
can't handle spaces in function names in XROM definition files.  
The 'geese' are displayed as the corresponding punctuation characters.
.SH AUTHOR
.B rom41cat
was written by Tony Duell, ard@p850ug1.demon.co.uk and has been placed 
under the GNU Public License version 2.0.


NEWS
change log information is in debian/changelog


src/lib/rom41cat.c
/* rom41cat.c -- print a catalogue of the functions in an unscrambled 
                 HP41 ROM dump */
/* 2001 A. R. Duell, and placed under the GPL */

#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <fcntl.h>
#include "xrom.h"
#include "config.h"
#include "lifutils.h"

/* Each 4K block of an HP41 ROM module starts with the FAT (Function Address
   Table). This contains the entry points for the functions in the ROM, 
   and whether they are in Focal (user code) or Mcode (machine code). 
   For machine code functions, the words preceding the entry point
   are the function name, using the HP41 'display' character set. For
   user code functions, the entry point is at a global label instruction, 
   the text of which is the name of the function.

   Further details of the format of the FAT (and of HP41 ROMs in general) 
   can be found in 'HP-41 Mcode for Beginners' by Ken Emery */

  
unsigned int rom41cat_read_rom(FILE *fp,unsigned int* rom)
/* Read a ROM image from fp */
  {
    unsigned int address; /* current ROM image address */ 
    unsigned char word[2]; /* 2 bytes for each 10 bit word */

    address=0;
    /* read in the ROM words */
    while(fread(word, sizeof(unsigned char), 2, fp)==2)
      {
         rom[address++]=(word[0]<<8)+word[1]; /* store the word */
      }
    return(address); /* And return the number of words read */
  }

unsigned char rom41cat_disp2asc (unsigned char disp_char)
/* Attempt to translate HP41 'display' characters into ascii */
  {
    /* Minor 'bug' : The geese are displayed as the corresponding punctuation
       characters */
    unsigned char row4[16] = 
    {127,'a','b','c','d','e',0,96,6,4,5,1,12,29,126,13};
    if (disp_char>79) { return('?');} /* illegal char */
    if (disp_char<32) { return(disp_char+64); } /* upper case ascii */
    if (disp_char<64) { return(disp_char); } /* digits and punctuation */
    return(row4[disp_char-64]); /* misc characters */
  }


void rom41cat_display_names(unsigned int * rom, unsigned int rom_size, int xrom_flag, int utf_flag)
/* Go through the FAT and print out the names */
  {
    unsigned int xrom; /* XROM id for this page */
    unsigned int n_funcs; /* number of functions in this 4K page */
    unsigned int func; /* current function number */
    unsigned int n_pages; /* number of 4K pages in this image */
    unsigned int page; /* current page */ 
    unsigned int fat_address; /* address of FAT entry for this function */
    unsigned int start_address; /* start address of this function */
    unsigned int addr;
    unsigned int focal_flag; /* Is this function in focal? */
    unsigned char c;
    int l,i;

    unsigned char fname[MAX_ALPHA];
    
    n_pages = (rom_size+4095)/4096; /* number of 4K pages to look at */
    /* look at each page in turn */
    if(! xrom_flag) printf("Number of pages %d ROM size %d\n",n_pages, rom_size);
    for(page=0; page<n_pages; page++)
      {
        xrom = rom[page*4096]; /* XROM id number */
        n_funcs = rom[page*4096+1]; /* number of functions in the FAT */

        if(!xrom_flag) printf("Page %d, Number of Functions %d\n",page+1,n_funcs);
        for(func=0; func<n_funcs; func++)
          {
            fat_address=page*4096 + 2*func + 2;
            start_address=((rom[fat_address]&0x7f)<<8) 
                         +(rom[fat_address+1]&0xff)
                         +page*4096;
            // if the start address is outside the rom address space print message on stderr
            if(start_address> rom_size) {
                fprintf(stderr,"%d %d function addr %x outside of ROM address space\n",xrom,func,start_address);
                continue;
            }
            focal_flag=rom[fat_address]&0x200;
            if(focal_flag) {
               l=rom[start_address+2]-0xF1; /* get the string length nybble */
               for(i=0;i<l;i++) {
                  c= rom[start_address+i+4]&0x7f;
                  fname[i]=((c==' ')?'_':c);
               }
            } else {
               l=0;
               addr= start_address;
               do {
                  c=rom41cat_disp2asc((rom[--addr]&0x7f));
                  fname[l]=((c==' ')?'_':c);
                  l++;
               }
               while(!(rom[addr]&0x80));
            }
            if(xrom_flag) {
               printf("%d %d ",xrom,func);
               if(focal_flag) {
                   printf("XROM'%s'",to_hp41_string(fname,l,0));
                   if(utf_flag && has_special_characters(fname,l)) {
                      printf(" XROM'%s'",to_hp41_string(fname,l,1));
                   }
               } else {
                   printf("%s",to_hp41_string(fname,l,0));
                   if(utf_flag && has_special_characters(fname,l)) {
                      printf(" %s",to_hp41_string(fname,l,1));
                   }
               }
               printf("\n");
           } else {
              printf("XROM %02d,%02d ",xrom,func);
              /* for user output, print the ID, language and entry point */
              printf("Entry = %04x ", start_address);
              printf("(%s) ",focal_flag?"Focal":"Mcode");
              printf("%s\n",to_hp41_string(fname,l,utf_flag));
           }
        }
      }
  }

void rom41cat_usage(void)
  {
    fprintf(stderr,"Usage : lifutils rom41cat [-x] < input file [> output file]\n");
    fprintf(stderr,"        or\n");
    fprintf(stderr,"        lifutils rom41cat [-x] INPUTFILE [> output file]\n");
    fprintf(stderr,"        Display the function names in a HP41 ROM image\n");
    fprintf(stderr,"        -x : output in XROM file format\n");
    fprintf(stderr,"\n");
  }

int rom41cat(int argc, char **argv)
  {
    unsigned int rom_size; /* size of input image */
    int xrom_flag=0; /* output in XROM file format? */
    int utf_flag=0;  /* output names with utf characters */
    int option; /* current option character */
    FILE *fp;

    /* Buffer to store ROM image */
    unsigned int rom[16384];


    /* Process command line options */
    optind=1;
    while((option=getopt(argc,argv,"ax?"))!=-1)
      {
        switch(option)
          {
            case 'x' : xrom_flag=1;
                       break;
            case 'a' : utf_flag=1;
                       break;
            case '?' : rom41cat_usage();
                       return(RETURN_OK);
          }
      }

    if((optind!=argc) && (optind!= argc-1))
       {
         rom41cat_usage();
         return(RETURN_ERROR);
       }
    if(optind== argc-1) {
        fp=fopen(argv[argc-1],"rb");
     } else {
        fp=stdin;
        SETMODE_STDIN_BINARY;
     }

     if (fp == NULL) {
       fprintf(stderr,"Error: cannot open input file\n");
       return(RETURN_ERROR);
     }


    /* Read in the ROM */
    rom_size=rom41cat_read_rom(fp,rom);
    /* And print the function names */
    rom41cat_display_names(rom, rom_size,xrom_flag, utf_flag);
    if(fp!=stdin) fclose(fp);
    return(RETURN_OK);
  }

